Dev master candidate from NCAR - 2019 06 11

gustavo-marques opened the pull request:
The main goal of this PR is to bring the latest changes in dev/ncar into dev/master.  A summary of the main changes/capabilities contained in this PR is provided below.

### GME and GEOMETRIC

* Add GME backscatter (Bachman, 2019). MEKE must be enabled for this parameterization to be used, as it uses MEKE to limit the amount of energy it adds.
* Add GEOMETRIC (Mak et al., 2017). This is an add-on to the MEKE package and is selected as a runtime parameter. Presently the only difference between GEOMETRIC and MEKE is in the expression for the thickness diffusivity.
* Add biharmonic option for MEKE viscosity.
* Add ability to count GME as an energy sink for MEKE. This is switched off by default.
* Add ability to count a non-MEKE thickness diffusivity as an energy source for MEKE. Also added the ability to use this thickness diffusivity to diffuse MEKE (previously it could only be used with the diffusivity diagnosed within MEKE).
* Add a proper expression for the Jansen et al. (2015) bottom drag into MEKE.
* Add a sign-definite way of accounting for the GM source term in MEKE.
* Add a sign-definite way of accounting for the frictional (viscous) source term in MEKE.

### MCT and NUOPC caps

* Change in latent heat flux calculation. The latent heat used to be provided via coupler and is now computed in the caps (MCT and NUOPC) via fprec, rofi_flux and q_flux.

* Add missing fluxes. MOM6 (coupler) definition of these terms is seaice_melt (meltw) and seaice_melt_heat (melth).

* Fix ocean coupling lag in the NUOPC cap.

All code has been properly documented (doxygenized). 

This PR does not change answers for the GFDL tests based on 'dev-master-candidate-2019-04-26'. It also does not change answers for NCAR.

adcroft commented:
This is a PR onto dev/master from NCAR. @awallcraft, @kshedstrom, @jiandewang feedback please.

adcroft commented on src/parameterizations/lateral/MOM_thickness_diffuse.F90:
```diff
@@ -146,17 +162,22 @@ subroutine thickness_diffuse(h, uhtr, vhtr, tv, dt, G, GV, US, MEKE, VarMix, CDp
     cg1 => null()
   endif
 
+  epsilon = 1.e-6
```
This is a dimensional number so must have appropriate unit scaling. @Hallberg-NOAA  ro @marshallward can advise on how. I also think this should be a run-timeparameter since 1e-6 s-1 might not be the best number to use always.

adcroft replied:
@Hallberg-NOAA just pointed out that 1e-6 is a large number relative to SN, not small !?

adcroft commented:
Answers for MOM6-examples do not reproduce with this PR without intervention.

The following experiments in MOM6-examples need `USE_VISBECK=True`:
- coupled_AM2_LM3_SIS/CM2G63L
- coupled_AM2_LM3_SIS2/AM2_SIS2_MOM6i_1deg
- coupled_AM2_LM3_SIS2/Concurrent_ice_1deg
- ice_ocean_SIS2/Baltic
- ice_ocean_SIS2/SIS2
- ice_ocean_SIS2/SIS2_bergs_cgrid
- ice_ocean_SIS2/SIS2_cgrid
- ocean_only/benchmark
- ocean_only/nonBous_global

The default for AH_VEL_SCALE was made non-zero. This causes answers changes in addition to the above. The default for this should be 0.

Parameter MEKE_VISCOSITY_COEFF was renamed to MEKE_VISCOSITY_COEFF_KU and needs to be set in ice_ocean_SIS2/OM4_05 . We should obsolete the old name if we are to change parameter names.

jiandewang commented:
I cloned MOM6 but I don't see this branch, how do I checkout them ?

On Mon, Jun 17, 2019 at 11:22 AM Alistair Adcroft <notifications@github.com>
wrote:

> This is a PR onto dev/master from NCAR. @awallcraft
> <https://github.com/awallcraft>, @kshedstrom
> <https://github.com/kshedstrom>, @jiandewang
> <https://github.com/jiandewang> feedback please.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/MOM6/pull/935?email_source=notifications&email_token=ADMGAS2CEJJGAG744TF45I3P26T3RA5CNFSM4HXEEU42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX3QT7Q#issuecomment-502729214>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADMGAS3UWZYS5N32BP4XC3TP26T3RANCNFSM4HXEEU4Q>
> .
>

gustavo-marques commented:
@jiandewang, here is the branch:
 https://github.com/gustavo-marques/MOM6/tree/dev-master-candidate-2019-06-11

kshedstrom commented:
This code changes answers for me in Tidal_bay.

This code causes CCS1 to somehow think the date is different from before and it now fails with:

`FATAL from PE    34: time_interp_external 2: time 711141 (19480101.090000 is before range of list 722829-723194(19800101.000000 - 19801231.210000),file=/center1/AKWATERS/kate/JRA55-do/...`

kshedstrom commented:
Actually, it's still a problem in my CCS1 input files, not the code change. I have work to do...

jiandewang commented:
passed compiling step on NCEP theia machine in EMC Unified Forecast System (UFS), more testing is ongoing

jiandewang commented:
passed compiling on EMC theia and wcoss machine

gustavo-marques commented:
> Answers for MOM6-examples do not reproduce with this PR without intervention.
> 
> The following experiments in MOM6-examples need `USE_VISBECK=True`:
> 
> * coupled_AM2_LM3_SIS/CM2G63L
> * coupled_AM2_LM3_SIS2/AM2_SIS2_MOM6i_1deg
> * coupled_AM2_LM3_SIS2/Concurrent_ice_1deg
> * ice_ocean_SIS2/Baltic
> * ice_ocean_SIS2/SIS2
> * ice_ocean_SIS2/SIS2_bergs_cgrid
> * ice_ocean_SIS2/SIS2_cgrid
> * ocean_only/benchmark
> * ocean_only/nonBous_global
> 
> The default for AH_VEL_SCALE was made non-zero. This causes answers changes in addition to the above. The default for this should be 0.
> 
> Parameter MEKE_VISCOSITY_COEFF was renamed to MEKE_VISCOSITY_COEFF_KU and needs to be set in ice_ocean_SIS2/OM4_05 . We should obsolete the old name if we are to change parameter names.

Just to confirm, GFDL will set `USE_VISBECK=True` in the above experiments to avoid change in answers. Correct?

adcroft commented:
@gustavo-marques Correct. We'll add `USE_VISBECK=True` where necessary.

gustavo-marques commented:
@septicscuzzy and I have fixed a few of the issues mentioned above. Here is a summary of the latest commits:

* Obsoleted MEKE_VISCOSITY_COEFF;
* Created a run-time parameter MEKE_GEOMETRIC_EPSILON and deleted hard-coded epsilon;
* Moved OMP calls to after GME and before the k-loop in MOM_hor_visc;
* Documented  MEKE_VISCOSITY_COEFF_AU and MEKE_VISCOSITY_COEFF_KU in the MEKE module.

**TO DO:**

1) We have not commented out the calls to the barotropic module since doing so will break GME for everybody. In the future, these calls will be replaced by a generic function that returns a barotropic velocity outside the barotropic control structure;
2) Unit-scaling will be added in the future.

kshedstrom commented:
Tidal_bay is changing in the horizontal viscosity. MOM_hor_visc.F90 has changed a great deal and totalview is showing me the wrong version in one of my dualling sessions, but I do know that diffu and diffv are different after the first call to horizontal_viscosity.

I don't remember why it is this way exactly, but the options in MOM_input are:
```
! === module MOM_hor_visc ===
LAPLACIAN = True                !   [Boolean] default = False
                                ! If true, use a Laplacian horizontal viscosity.
BIHARMONIC = True               !   [Boolean] default = True
                                ! If true, se a biharmonic horizontal viscosity.
                                ! BIHARMONIC may be used with LAPLACIAN.
KH = 0.0E+02                    !   [m2 s-1] default = 0.0
                                ! The background Laplacian horizontal viscosity.
KH_VEL_SCALE = 0.003            !   [m s-1] default = 0.0
                                ! The velocity scale which is multiplied by the grid
                                ! spacing to calculate the Laplacian viscosity.
                                ! The final viscosity is the largest of this scaled
                                ! viscosity, the Smagorinsky viscosity and KH.
SMAGORINSKY_KH = True           !   [Boolean] default = False
                                ! If true, use a Smagorinsky nonlinear eddy viscosity.
SMAG_LAP_CONST = 0.15           !   [nondim] default = 0.0
                                ! The nondimensional Laplacian Smagorinsky constant,
                                ! often 0.15.
SMAGORINSKY_AH = True
SMAG_BI_CONST = 0.6
```

adcroft commented:
Did you try AH_VEL_SCALE=0. ?--Dr Alistair Adcroft    (Alistair.Adcroft@noaa.gov)Princeton University           Tel: (609) 987-5073NOAA/GFDL, 201 Forrestal Road, Princeton, NJ 08540On Jun 21, 2019 20:02, Kate Hedstrom <notifications@github.com> wrote:Tidal_bay is changing in the horizontal viscosity. MOM_hor_visc.F90 has changed a great deal and totalview is showing me the wrong version in one of my dualling sessions, but I do know that diffu and diffv are different after the first call to horizontal_viscosity.
I don't remember why it is this way exactly, but the options in MOM_input are:
! === module MOM_hor_visc ===
LAPLACIAN = True                !   [Boolean] default = False
                                ! If true, use a Laplacian horizontal viscosity.
BIHARMONIC = True               !   [Boolean] default = True
                                ! If true, se a biharmonic horizontal viscosity.
                                ! BIHARMONIC may be used with LAPLACIAN.
KH = 0.0E+02                    !   [m2 s-1] default = 0.0
                                ! The background Laplacian horizontal viscosity.
KH_VEL_SCALE = 0.003            !   [m s-1] default = 0.0
                                ! The velocity scale which is multiplied by the grid
                                ! spacing to calculate the Laplacian viscosity.
                                ! The final viscosity is the largest of this scaled
                                ! viscosity, the Smagorinsky viscosity and KH.
SMAGORINSKY_KH = True           !   [Boolean] default = False
                                ! If true, use a Smagorinsky nonlinear eddy viscosity.
SMAG_LAP_CONST = 0.15           !   [nondim] default = 0.0
                                ! The nondimensional Laplacian Smagorinsky constant,
                                ! often 0.15.
SMAGORINSKY_AH = True
SMAG_BI_CONST = 0.6


—You are receiving this because you commented.Reply to this email directly, view it on GitHub, or mute the thread.

kshedstrom commented:
With AH_VEL_SCALE = 0, both executables give the same answer - and the same answer as I was getting from dev/esmg (dev/gfdl).

Hallberg-NOAA commented:
I think that the default value for AH_VEL_SCALE should be set back to 0 in the MOM_hor_visc code.  We have no particular reason to choose one dimensional value (0.10 m s-1) over any other. Where we set a default dimensional value, it is usually based on the well-defined properties of planet Earth (like its rotation rate, gravitational acceleration and radius) or something to do with the observable properties of sea-water (like its heat capacity).

adcroft commented:
@gustavo-marques Can you please change the default for AH_VEL_SCALE back to 0. ?

gustavo-marques commented:
> @gustavo-marques Can you please change the default for AH_VEL_SCALE back to 0. ?

Done.

kshedstrom commented:
I now approve this PR.

jiandewang commented:
I am out of town now, will revisit this PR next Monday when I am back.

On Wed, Jun 26, 2019 at 7:06 PM Kate Hedstrom <notifications@github.com>
wrote:

> I now approve this PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/MOM6/pull/935?email_source=notifications&email_token=ADMGAS4GQQUCB3MAG5YJANLP4PZABA5CNFSM4HXEEU42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYVB4IY#issuecomment-506076707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADMGAS66J5RFGLP6OH3NS23P4PZABANCNFSM4HXEEU4Q>
> .
>

Hallberg-NOAA commented:
I continue to have strong misgivings about the addition of the barotropic_CS and thickness_diffuse_CS as non-optional arguments to horizontal_viscosity.  These additions required substantial changes in the arguments to the overlying code, and they will not work anyway unless the split time stepping scheme is used.  However, I think that I have a workable solution to these two pieces that might help us get this code merged in a way that we will not regret later.

1. The thickness_diffuse_CS is not actually being used in horizontal_viscosity, so please just remove it as an argument to horizontal_viscosity and the 3 stepMOM_dyn routines. We can collectively come back and put some effort into thinking about how to properly handle sharing information between the horizontal viscosity and GM diffusivity (misnamed thickness-diffuse), but for now there is no reason for this change.

2. Make barotropic an _optional_ argument to horizontal_viscosity that is only passed in with the SPLIT=True time stepping scheme, and then add an error message to trap the case if anyone tries to set USE_GME=True when this argument is missing.  Currently, I am pretty sure that the barotropic_CS is not even being allocated when SPLIT=False, and trying to use USE_GME=True with SPLIT=False will lead to a segmentation fault without any useful error message.  This approach will keep USE_GME working as it currently does in any case that works now, and it improves the error handling and will help lay the groundwork to come up with a different way to get at the barotropic velocities in the un-split cases.  This solution will then allow us to remove the unallocated barotropic_CS argument from step_MOM_dyn_unsplit and step_MOM_dyn_unsplit_RK2, which would then be identical to before, while the solutions will be identical in all of your new cases that use GME.

  @gustavo-marques and @septicscuzzy, please let me know if you think that this approach is reasonable. If so, I can help you with these changes as a correction to your PR.

gustavo-marques commented:
@Hallberg-NOAA, thanks for proving a solution to these issues. @septicscuzzy and I like your idea. I will start working on it today and will let you know if I have questions.

gustavo-marques commented:
This PR has been updated following @Hallberg-NOAA's idea for prohibiting USE_GME=True when SPLIT=False while keeping GME's functionality when SPLIT=True. 

Unfortunately, we are unable to fully test these changes at NCAR -- Cheyenne has been down since last week and there was a problem rebooting the system this morning. Instead of waiting for Cheyenne to be fixed, I've decided to update this PR anyway so others can review and test these changes.

jiandewang commented:
works fine on EMC theia and wcoss machines


On Mon, Jul 1, 2019 at 11:57 AM Gustavo Marques <notifications@github.com>
wrote:

> This PR has been updated following @Hallberg-NOAA
> <https://github.com/Hallberg-NOAA>'s idea for prohibiting USE_GME=True
> when SPLIT=False while keeping GME's functionality when SPLIT=True.
>
> Unfortunately, we are unable to fully test these changes at NCAR --
> Cheyenne has been down since last week and there was a problem rebooting
> the system this morning. Instead of waiting for Cheyenne to be fixed, I've
> decided to update this PR anyway so others can review and test these
> changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-GFDL/MOM6/pull/935?email_source=notifications&email_token=ADMGASY6SQWRB3J6ITEHD4LP5ISONA5CNFSM4HXEEU42YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY6SSAA#issuecomment-507324672>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADMGAS43DH4SMZZGRYSJK63P5ISONANCNFSM4HXEEU4Q>
> .
>

adcroft commented:
I've just retested this PR again up to commit 5c42192ffe66122 and we're good-to-go with the caveat that USE_VISBECK needs to be added to particular experiments. So we are waiting for NCAR (@gustavo-marques) to confirm their tests.

gustavo-marques commented:
It passes our tests (b4b on C- and G-compsets using tx06).
